webxrlayers-1.bs

```diff
@@ -912,63 +915,50 @@ WebXR Device API Integration {#webxrintegration}
-  1. Initialize |state|'s {{XRRenderState/layers}} to an empty array.
+This module replaces the steps given by [=initialize the render state=]. Instead when an {{XRRenderState}} object |state|
+is created for an {{XRSession}} |session|, the user agent MUST <dfn dfn-for="layers">initialize the render state</dfn> by running the following steps:
+  1. Initialize |state| by running the original steps to [=initialize the render state=].
+  1. Initialize |state|'s {{XRRenderState/layers}} as follows:
+    <dl class="switch">
+      <dt class="switch">If the |session| was created with "[=feature descriptor/layers=]" enabled
+      <dd> Initialize |state|'s {{XRRenderState/layers}} with a new empty array.
+      <dt> Otherwise
+      <dd> set |state|'s {{XRRenderState/layers}} to <code>null</code>.
+    </dl>
 
 </div>
 
 updateRenderState changes {#updaterenderstatechanges}
 -------------------------
-<div class="algorithm" data-algorithm="updateRenderState-for-layers">
-When the <dfn method for="XRSession" dfn-for="layers">updateRenderState(|newState|)</dfn> method is called on an {{XRSession}} |session|
-with "[=feature descriptor/layers=]" enabled, the user agent MUST run the following steps:
 
-  1. Run {{XRSession/updateRenderState}} with |newState|.
-  1. If |newState|'s [=XRRenderStateInit/baseLayer=] is set, set |session|'s [=pending render state=] to <code>null</code>, throw an {{InvalidStateError}} and abort these steps.
+<div class="algorithm" data-algorithm="update-layers-state-for-layers">
+
+This module replaces the steps given by "[=update the pending layers state=]". Instead when the user agent will
+<dfn dfn-for="layers">update the pending layers state</dfn> with {{XRSession}} |session| and {{XRRenderStateInit}} |newState|, it must run the following steps:
+  1. If the |session| was not created with "[=feature descriptor/layers=]" enabled:
+    1. If {{XRRenderStateInit/layers}} is not <code>null</code>, throw an {{NotSupportedError}} and abort these steps.
+    1. Abort these steps.
+  1. If |newState|'s {{XRRenderStateInit/baseLayer}} is set, throw a {{InvalidStateError}} and abort these steps.
+  1. If |newState|'s {{XRRenderStateInit/layers}} contains duplicate instances, throw an {{InvalidStateError}} and abort these steps.
   1. For each |layer| in |newState|'s {{XRRenderStateInit/layers}}:
-    1. If |layer|'s [=XRLayer/session=] is different from |session|, throw an {{InvalidStateError}} and abort these steps.
+    1. If |layer| is a {{XRGenericLayer}} and |layer|'s [=XRGenericLayer/session=] is different from |session|, throw an {{InvalidStateError}} and abort these steps.
+    1. If |layer| is a {{XRWebGLLayer}} and |layer|'s [=XRWebGLLayer/session=] is different from |session|, throw an {{InvalidStateError}} and abort these steps.
   1. If |newState|'s {{XRRenderStateInit/layers}} array is set, set |session|'s [=pending render state=]'s {{XRRenderState/layers}} to |newState|'s {{XRRenderStateInit/layers}}.
```

Comment:
This should also set `pending/baseLayer` to null, and we should have another line that nulls out `pending/layers` when `baseLayer` is set.

Reply:
Are you saying layers shouldn't be null?
Like: If baseLayers is set or layers is null, throw InvalidStateError

I think that would be wrong.

Reply:
Why is that needed?

Reply:
Otherwise you can do `updateRenderState(baseLayer: foo)` followed by `updateRenderState(layers: bar)`. We either need to silently null out `baseLayer` when `layers` is set and vice versa, or we need to have code that ensure that the active render state also does not have baseLayer or layers set when we are trying to update the other without explicitly nulling out the first.

Reply:
ah. But you can't do that anyway :-)
If the layers feature isn't set, you will error out if you pass a layers array
Likewise, if it is set, you will get an error if you pass a baseLayer.

Reply:
oh! I didn't realize that, that's clever

I don't see where in the algorithm we error if baseLayer is used with the feature set, though?

And this doesn't handle cases where the feature set is appended midway

Reply:
> I don't see where in the algorithm we error if baseLayer is used with the feature set, though?

See line 954. It's hard to see on the review.

> And this doesn't handle cases where the feature set is appended midway

That can't happen. The feature is requested when the session is created.

Reply:
> That can't happen. The feature is requested when the session is created.

No, with the permissions API integration (I don't think this is shipped anywhere yet, but it's there), you can request features at arbitrary points

https://immersive-web.github.io/webxr/#ref-for-xr-device-list-of-enabled-features%E2%91%A0

You could declare the layers feature as one which _has_ to be requested at session creation. @toji ?

Reply:
> You could declare the layers feature as one which _has_ to be requested at session creation. @toji ?

I think it doesn't make sense to add this at runtime because an author can use the classic welgllayer as the sole entry in the array. Granting this permission will allso put the renderState in a bad state.

Reply:
> Granting this permission will allso put the renderState in a bad state.

Right, that's why I was initially suggesting the additional check.

We should specify somewhere that the "layers" feature cannot be requested after session creation.

Reply:
Hrm. I forgot about the permissions API's mid-session feature request capability.

I do think there are going to be features that don't make sense to request once the session has begun, largely if the underlying runtime requires that the feature be enabled at startup. In this particular I don't think that restriction applies, but I also can't think of any reason why the page would want to switch layering style mid-session, so I woud have no problem with putting that restriction in place. I don't know what the best way to flag that would be, though.

Reply:
I will file an issue on this so we're not gated on solving it in the PR.